If archive have a new name, try to get it from Github release page

Yannig opened the pull request:
## Description

If you try to download kubeseal with arkde 0.8.12, you get v0.14.1. Now trying to get v0.17.1, you get an error:

```
arkade get kubeseal@v0.17.1
```

```
Downloading: kubeseal
Downloading: https://github.com/bitnami-labs/sealed-secrets/releases/download/v0.17.1/kubeseal-linux-amd64
Error: incorrect status for downloading tool: 404
```

## Motivation and Context
- [X] I have raised an issue to propose this change, which has been given a label of `design/approved` by a maintainer ([required](https://github.com/alexellis/arkade/blob/master/CONTRIBUTING.md))


## How Has This Been Tested?

Download kubeseal version v0.14.1 (direct download of the binary):

```
arkade get kubeseal@v0.14.1
Downloading: kubeseal
Downloading: https://github.com/bitnami-labs/sealed-secrets/releases/download/v0.14.1/kubeseal-linux-amd64
33.56 MiB / 33.56 MiB [----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------] 100.00%
/tmp/kubeseal-linux-amd64 written.
...
```

```
file /home/yannig/.arkade/bin/kubeseal
/home/yannig/.arkade/bin/kubeseal: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), statically linked, Go BuildID=uDciUD6TUaSCxCVMfL3S/R9gWKA3w-zpjU2-Sy7KK/jwzKqQMM1X-dRjEmggjY/cW09PaeFey-p85_XeuNW, not stripped
```

Now download version v0.17.1 (embedded inside a tar.gz archive):
```
arkade get kubeseal@v0.17.1

Downloading: kubeseal
Downloading: https://github.com/bitnami-labs/sealed-secrets/releases/download/v0.17.1/kubeseal-linux-amd64
Error while downloading template URL (https://github.com/bitnami-labs/sealed-secrets/releases/download/v0.17.1/kubeseal-linux-amd64)
Try to autodetect archive url in Github release
Archive /bitnami-labs/sealed-secrets/releases/download/v0.17.1/kubeseal-0.17.1-linux-amd64.tar.gz seems to be compatible with linux/x86_64
Autodetect archive name inside Github: https://github.com/bitnami-labs/sealed-secrets/releases/download/v0.17.1/kubeseal-0.17.1-linux-amd64.tar.gz
13.97 MiB / 13.97 MiB [----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------] 100.00%
/tmp/kubeseal-0.17.1-linux-amd64.tar.gz written.
/tmp/LICENSE LICENSE
/tmp/README.md README.md
/tmp/kubeseal kubeseal
2022/01/20 14:48:06 extracted tarball into /tmp: 3 files, 0 dirs (252.940385ms)
2022/01/20 14:48:06 Extracted: /tmp/kubeseal
2022/01/20 14:48:06 Copying /tmp/kubeseal to /home/yannig/.arkade/bin/kubeseal

Tool written to: /home/yannig/.arkade/bin/kubeseal
...
```

```
file /home/yannig/.arkade/bin/kubeseal
/home/yannig/.arkade/bin/kubeseal: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), statically linked, Go BuildID=3tbTRrSgy9uJfZVH27iS/uugQD8256WqGh51DVHiq/FE2T9o6gXx--tLPAxez6/zTb2jyY8RMvwQ2pxJytN, not stripped
```

You can download both version as you want.

In reality, these new functions should make it possible to avoid some declaration for simple cases. The OS/Arch auto-detection feature relies on the contents of the release page.

## Are you a GitHub Sponsor (Yes/No?)

- [ ] Yes
- [X] No

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Documentation

- [ ] I have updated the list of tools in README.md if (required) with `./arkade get -o markdown`
- [ ] I have updated the list of apps in README.md if (required) with `./arkade install --help`

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I've read the [CONTRIBUTION](https://github.com/alexellis/arkade/blob/master/CONTRIBUTING.md) guide
- [X] I have signed-off my commits with `git commit -s`
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

<!--- In case it is a new application -->
- [ ] I have tested this on arm, or have added code to prevent deployment

alexellis commented:
Looks like we have merge conflicts that need to be fixed before I can look at this for you. 

Do you also mind removing the emojis from your commit message? Happy for you to put them in your pull request description instead.

alexellis commented:
Wow there are so many changes. Why?

Yannig commented:
I am aware that there are indeed a lot of changes but for the most part it is the result of two things:
- Addition of detection of systems or processor architectures (isOsCompatible, isArchCompatible) and the capacity to parse available files in an existing Github release (getGithubReleaseArchive). Not couting the corresponding unit test
- the addition of the version in the results returned by the functions GetDownloadURL, GetURL and getByDownloadTemplate.

These changes are necessary in order to propagate the version detected during calls to the Download function. Unfortunately, the change of GetURL has the consequence of changing a lot of stuff in the unit tests.

I will try to rework my PR to reduce these changes.

Yannig commented:
I reworked my commit to reduce the impact on existing code.

Essentially, I moved the call to the getToolVersion functions as well as IsArchive in order to simplify things.

The rest consists of the addition of search functions on the type of OS (isOsCompatible) or arch (isArchCompatible). These two functions are then used by getGithubReleaseArchive. This function retrieve files from given release and keep only the one compatible with the current request.

The rest of the commit consists of adding unit tests.

alexellis commented:
Please see #611 which is more like what I was expecting. Let us know if you see any differences in the final result.
